### DISCUSSION TITLE
is it possible to defer `pytest_addoption` in internal plugins?

### DISCUSSION THREAD

**Discussion by symonk** [pytest-dev/pytest#9971] [Q&A]
Creating a batch of internal plugins is a common use case; these are registered internally using `pytest_plugins = (...)` in a top level `conftest.py` however when performing the following, is it possible to have `pytest_addoption` invoked when a class is registered without complaining that no such arguments exist during CLI invocation.  I think this is maybe called very early on during config preperation and this simple isn't possible? (they should live in the module as a function instead):

```python
# example.py

def pytest_configure(config):
    config.pluginmanager.register(Foo())

class Foo:

    def pytest_addoption(parser):
        # add some options
```

Consider the above pseudo code, writing it from memory.  Is it possible that Foo being registered here inside example.py where `conftest.py` contains a `pytest_plugins = ['some.path.to.example']` can allow the added options to be usable without having them be strictly set in `example.py` as a function hook implementation of `pytest_addoption`.

Use case: I have a `meta` plugin that is responsible for registering various internal plugins, each of which have a lengthy list of options; I don't want to drag all the options into `example.py` here just for them to work.

Thanks!

**Comment by nicoddemus** [comment]
>  I think this is maybe called very early on during config preperation and this simple isn't possible?

I'm pretty sure that's the case, as it stands today.

**Comment by RonnyPfannschmidt** [comment]
Pytest configure is too late

Your use case sounds like a horror for ux, if only scripts fill in those options, consider confi files
